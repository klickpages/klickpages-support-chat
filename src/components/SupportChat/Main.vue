<template>
  <div id="support-chat"></div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import { setklickartURL } from '@/config/klickart';

export default {
  name: 'SupportChat',
  props: {
    klickartURL: String,
  },
  computed: {
    ...mapGetters({
      chat: 'SupportChat/chat',
    }),

    hasSupportChat() {
      const chatWrapperClass = 'md-chat-widget-wrapper';
      return !!document.body.querySelector(`.${chatWrapperClass}`);
    },
  },
  methods: mapActions({
    getChat: 'SupportChat/getChat',
  }),
  async mounted() {
    setklickartURL(this.klickartURL);

    if (this.hasSupportChat) {
      return;
    }

    try {
      const chatScript = await this.getChat();

      const HTMLChatFragment = document
        .createRange()
        .createContextualFragment(chatScript);

      this.$el.appendChild(HTMLChatFragment);
    } catch (error) {
      this.$notify.error({
        message: this.$t('supportChat.errorMessage'),
        autoHide: true,
      });
    }
  },
};
</script>
