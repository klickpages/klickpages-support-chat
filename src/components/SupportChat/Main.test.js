import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import SupportChat from './Main.vue';

const $notify = {
  error: jest.fn(),
  success: jest.fn(),
};

const $t = jest.fn().mockImplementation((str) => str);

describe('@/components/SupportChat/Main', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;

  const modules = {
    SupportChat: {
      actions: {
        getChat: jest.fn(),
      },
      namespaced: true,
    },
  };

  describe('when component is mounted', () => {
    describe('when there is no support chat in the page', () => {
      beforeAll(() => {
        const store = new Vuex.Store({
          modules,
        });

        wrapper = shallowMount(SupportChat, {
          localVue,
          store,
          mocks: { $t, $notify },
        });
      });

      afterAll(() => {
        wrapper.destroy();
        jest.clearAllMocks();
      });

      it('should render html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('when there is a support chat in the page', () => {
      beforeAll(() => {
        const mockReturnData = {
          data: '<script>sample data</script>',
        };

        modules.SupportChat.actions = {
          getChat: jest.fn().mockImplementation(() => Promise.resolve(mockReturnData)),
        };

        const store = new Vuex.Store({
          modules,
        });

        wrapper = shallowMount(SupportChat, {
          localVue,
          store,
          mocks: { $t, $notify },
        });
      });

      afterAll(() => {
        wrapper.destroy();
        jest.clearAllMocks();
      });

      it('should render html correctly', () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });

    describe('when there is no support chat in the page', () => {
      beforeAll(() => {
        modules.SupportChat.actions = {
          getChat: jest.fn().mockImplementation(() => Promise.reject()),
        };

        const store = new Vuex.Store({
          modules,
        });

        wrapper = shallowMount(SupportChat, {
          localVue,
          store,
          mocks: { $t, $notify },
        });
      });

      afterAll(() => {
        wrapper.destroy();
        jest.clearAllMocks();
      });

      it('should call Notify.error', () => {
        expect($notify.error).toHaveBeenCalledWith({
          message: 'Error message',
          autoHide: true,
        });
      });
    });
  });
});
