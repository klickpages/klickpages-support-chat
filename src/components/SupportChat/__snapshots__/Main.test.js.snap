// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/SupportChat/Main when component is mounted when there is a support chat in the page should render html correctly 1`] = `
<div
  id="support-chat"
/>
`;

exports[`@/components/SupportChat/Main when component is mounted when there is no support chat in the page should render html correctly 1`] = `
<div
  id="support-chat"
/>
`;
